File: testdata/fixtures/java/vulnerable/ZipSlip.java
import java.io.*;
import java.util.zip.*;

public class ZipSlip {
    public void extractZip(String zipFile, String destDir) throws Exception {
        ZipInputStream zis = new ZipInputStream(new FileInputStream(zipFile));
        ZipEntry entry;
        while ((entry = zis.getNextEntry()) != null) {
            // VULNERABLE: using entry.getName() directly in new File without validation
            File destFile = new File(destDir, entry.getName());
            destFile.getParentFile().mkdirs();

            FileOutputStream fos = new FileOutputStream(destFile);
            byte[] buffer = new byte[1024];
            int len;
            while ((len = zis.read(buffer)) > 0) {
                fos.write(buffer, 0, len);
            }
            fos.close();
        }
        zis.close();
    }
}

File: testdata/fixtures/javascript/vulnerable/zip_slip.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yauzl from 'yauzl';

function extractZip(zipPath: string, destDir: string) {
  yauzl.open(zipPath, { lazyEntries: true }, (err, zipfile) => {
    zipfile.readEntry();
    zipfile.on('entry', (entry) => {
      // VULNERABLE: using entry.fileName directly in path.join without validation
      const destPath = path.join(destDir, entry.fileName);
      const writeStream = fs.createWriteStream(destPath);
      zipfile.openReadStream(entry, (err, readStream) => {
        readStream.pipe(writeStream);
      });
      zipfile.readEntry();
    });
  });
}
